### PR TITLE
add proxy download and filename encode convert

### PR DESCRIPTION
**Pull request by serendipity-crypto**
## Types of changes

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Code cleanup / Refactoring

## Description

First, I add url encode convert when a filename is constructed from url. For example, this will convert "%20" into " ".

image

Second, I add the fuction "async fn download_with_proxy". This will allow user easily download file through proxy.

image

## Checklist:

- [] I have updated the documentation accordingly
- [] I have updated the Changelog (if applicable)

Fixes: PeopleForBikes/PeopleForBikes.github.io#

### CODE REVIEW

File: src/downloader.rs [hunk 1]
```diff
@@ -45,10 +45,35 @@ impl Downloader {
     const DEFAULT_CONCURRENT_DOWNLOADS: usize = 32;
 
     /// Starts the downloads.
+    #[inline]
```
Comment (rgreinho) [inline src/downloader.rs:48]: Quick question: what is the benefit of inlining this function?

**Comment by rgreinho** [comment]
Thanks for your contribution @serendipity-crypto. I will merge it and add a bit of docs, then I'll release a new version of the crate.

Just a small question, more for me learning: what are the benefits of inlining the functions you marked as `#[inline]`?

**Comment by serendipity-crypto** [comment]
Originally, I was struggling with whether to rewrite a "download_with_proxy" or pull out the code that shared with "download". In the end I chose the latter. At the same time, "download" is a very commonly used function for this crate, so I added the #[inline] attribute so that the compiler can automatically expand into the internal code of "download" where the "download" is called, so that the function call is one layer less. In this way, the user calls "download", and the compiler expands directly in the same place to "self.download_inner (downloads, None).await".
This is my understanding of #[inline].

**Comment by serendipity-crypto** [comment]
But I just googled "rust inline". I found that only when link-time optimization is enable, functions would be inlined across crates. Maybe rewrite a "download_with_proxy" is a better choice.

**Comment by serendipity-crypto** [comment]
I remove #[inline] and update a new commit. It seems that I haven't learned enough. 🤣

**Comment by rgreinho** [comment]
Oh, but now there is a lot of code duplication. Can't we keep what you wrote before, simply without the `#[inline]`?

**Comment by serendipity-crypto** [comment]
Ok, I make it.

**Comment by rgreinho** [comment]
Thank you 👍

**Pull request merged**